File: bin/src/sync/kuzu_ts/core/client/kuzu_native_client.py
#!/usr/bin/env python3
"""
KuzuDB Native Client for TypeScript Integration
ネイティブKuzuDBをTypeScriptから使用するためのクライアント

This script acts as a bridge between TypeScript and native KuzuDB,
handling DML execution and state queries through JSON-RPC style communication.
"""

import sys
import json
import os

# Try importing kuzu, but fall back to mock if not available
try:
    import kuzu
    
    def create_database(path=None):
        """Create database using direct kuzu import"""
        if path is None or path == ":memory:":
            return kuzu.Database(":memory:")
        else:
            return kuzu.Database(str(path))
    
    def create_connection(database):
        """Create connection using direct kuzu import"""
        return kuzu.Connection(database)
except ImportError:
    # Mock implementation for testing architecture
    print("WARNING: Using mock KuzuDB implementation", file=sys.stderr)
    
    class MockDatabase:
        def __init__(self, path):
            self.path = path
            self.data = {}
    
    class MockConnection:
        def __init__(self, db):
            self.db = db
            self.executed_queries = []
        
        def query(self, cypher, params=None):
            self.executed_queries.append((cypher, params))
            # Mock result
            class MockResult:
                def getAllObjects(self):
                    return []
                def getAll(self):
                    return []
            return MockResult()
    
    def create_database(path=None):
        return MockDatabase(path or ":memory:")
    
    def create_connection(database):
        return MockConnection(database)

import traceback

def is_error(result):
    """Check if result is an error"""
    return isinstance(result, dict) and "type" in result and result["type"].endswith("Error")

class KuzuNativeClient:
    def __init__(self):
        self.db = None
        self.conn = None
        self.initialized = False
        
    def initialize(self, path=None):
        """Initialize KuzuDB database and connection"""
        try:
            # Create in-memory database by default
            db_path = path if path else ":memory:"
            self.db = create_database(db_path)
            
            if is_error(self.db):
                return {"error": self.db}
            
            self.conn = create_connection(self.db)
            
            if is_error(self.conn):
                return {"error": self.conn}
            
            # Create default schema
            self._create_default_schema()
            
            self.initialized = True
            return {"success": True, "message": "Database initialized"}
            
        except Exception as e:
            return {"error": str(e), "traceback": traceback.format_exc()}
    
    def _create_default_schema(self):
        """Create default schema for sync application"""
        queries = [
            # User node table
            """CREATE NODE TABLE IF NOT EXISTS User(
                id STRING, 
                name STRING, 
                email STRING, 
                PRIMARY KEY(id)
            )""",
            
            # Post node table
            """CREATE NODE TABLE IF NOT EXISTS Post(
                id STRING,
                content STRING,
                authorId STRING,
                PRIMARY KEY(id)
            )""",
            
            # FOLLOWS relationship table
            """CREATE REL TABLE IF NOT EXISTS FOLLOWS(FROM User TO User)""",
            
            # Counter node table
            """CREATE NODE TABLE IF NOT EXISTS Counter(
                id STRING,
                value INT64,
                PRIMARY KEY(id)
            )"""
        ]
        
        for query in queries:
            result = self.conn.query(query)
            # Log creation
            self.log("INFO", {
                "message": "Schema created",
                "query": query.replace('\n', ' ').strip()
            })
    
    def execute_template(self, template, params):
        """Execute a DML template with parameters"""
        if not self.initialized:
            return {"error": "Client not initialized"}
        
        templates = {
            "CREATE_USER": """
                CREATE (u:User {
                    id: $id,
                    name: $name,
                    email: $email
                })
            """,
            "UPDATE_USER": """
                MATCH (u:User {id: $id})
                SET u.name = $name
            """,
            "CREATE_POST": """
                CREATE (p:Post {
                    id: $id,
                    content: $content,
                    authorId: $authorId
                })
            """,
            "FOLLOW_USER": """
                MATCH (follower:User {id: $followerId})
                MATCH (target:User {id: $targetId})
                CREATE (follower)-[:FOLLOWS]->(target)
            """,
            "INCREMENT_COUNTER": """
                MERGE (c:Counter {id: $counterId})
                ON CREATE SET c.value = COALESCE($amount, 1)
                ON MATCH SET c.value = c.value + COALESCE($amount, 1)
            """,
            "QUERY_COUNTER": """
                MATCH (c:Counter {id: $counterId})
                RETURN c.value as value
            """
        }
        
        query = templates.get(template)
        if not query:
            return {"error": f"Unknown template: {template}"}
        
        try:
            # Execute query
            result = self.conn.query(query, params)
            
            # Log execution
            self.log("INFO", {
                "message": "DML executed",
                "template": template,
                "params": params
            })
            
            # Handle query results
            if template == "QUERY_COUNTER":
                result_set = result.getAllObjects()
                if result_set:
                    return {"success": True, "result": result_set[0].get("value", 0)}
                else:
                    return {"success": True, "result": 0}
            
            return {"success": True}
            
        except Exception as e:
            self.log("ERROR", {
                "message": "DML execution failed",
                "template": template,
                "params": params,
                "error": str(e)
            })
            return {"error": str(e)}
    
    def get_local_state(self):
        """Get current database state"""
        if not self.initialized:
            return {"error": "Client not initialized"}
        
        try:
            # Get users
            user_result = self.conn.query("""
                MATCH (u:User)
                RETURN u.id as id, u.name as name, u.email as email
                ORDER BY u.id
            """)
            users = user_result.getAllObjects() if hasattr(user_result, 'getAllObjects') else []
            
            # Get posts
            post_result = self.conn.query("""
                MATCH (p:Post)
                RETURN p.id as id, p.content as content, p.authorId as authorId
                ORDER BY p.id
            """)
            posts = post_result.getAllObjects() if hasattr(post_result, 'getAllObjects') else []
            
            # Get follows
            follow_result = self.conn.query("""
                MATCH (follower:User)-[:FOLLOWS]->(target:User)
                RETURN follower.id as followerId, target.id as targetId
            """)
            follows = follow_result.getAllObjects() if hasattr(follow_result, 'getAllObjects') else []
            
            return {
                "success": True,
                "state": {
                    "users": users,
                    "posts": posts,
                    "follows": follows
                }
            }
            
        except Exception as e:
            return {"error": str(e)}
    
    def log(self, level, data):
        """Log to stderr for TypeScript telemetry"""
        log_entry = {
            "level": level,
            "timestamp": self._get_timestamp(),
            "uri": "sync/kuzu_ts/native_client",
            **data
        }
        print(json.dumps(log_entry), file=sys.stderr)
    
    def _get_timestamp(self):
        """Get ISO timestamp"""
        from datetime import datetime
        return datetime.now().isoformat()
    
    def execute_query(self, query, params=None):
        """Execute raw Cypher query"""
        if not self.initialized:
            return {"error": "Client not initialized"}
        
        try:
            result = self.conn.query(query, params or {})
            
            # Log execution
            self.log("DEBUG", {
                "message": "Raw query executed",
                "query": query.replace('\n', ' ').strip()
            })
            
            # Try to get results if available
            if hasattr(result, 'getAllObjects'):
                return {"success": True, "result": result.getAllObjects()}
            else:
                return {"success": True, "result": None}
            
        except Exception as e:
            self.log("ERROR", {
                "message": "Query execution failed",
                "query": query,
                "error": str(e)
            })
            return {"error": str(e)}
    
    def handle_request(self, request):
        """Handle JSON-RPC style request"""
        method = request.get("method")
        params = request.get("params", {})
        
        if method == "initialize":
            return self.initialize(params.get("path"))
        elif method == "executeTemplate":
            return self.execute_template(params["template"], params["params"])
        elif method == "getLocalState":
            return self.get_local_state()
        elif method == "executeQuery":
            return self.execute_query(params["query"], params.get("params"))
        else:
            return {"error": f"Unknown method: {method}"}


def main():
    """Main entry point - reads JSON requests from stdin, writes responses to stdout"""
    client = KuzuNativeClient()
    
    # Log startup
    client.log("INFO", {"message": "Native KuzuDB client started"})
    
    # Read line by line from stdin
    for line in sys.stdin:
        try:
            request = json.loads(line.strip())
            response = client.handle_request(request)
            print(json.dumps(response))
            sys.stdout.flush()
        except json.JSONDecodeError as e:
            error_response = {"error": f"Invalid JSON: {str(e)}"}
            print(json.dumps(error_response))
            sys.stdout.flush()
        except Exception as e:
            error_response = {"error": str(e), "traceback": traceback.format_exc()}
            print(json.dumps(error_response))
            sys.stdout.flush()


if __name__ == "__main__":
    main()

File: bin/src/sync/kuzu_ts/core/client/kuzu_native_client.ts
/**
 * KuzuDB Native Client - TypeScript Interface
 * Pythonネイティブクライアントへのインターフェース
 * 
 * Uses subprocess to communicate with Python KuzuDB client,
 * enabling actual DML execution unlike WASM implementation.
 */

import type { KuzuWasmClient, LocalState, EventSnapshot } from "../../types.ts";
import type { TemplateEvent } from "../../event_sourcing/types.ts";
import type { DDLTemplateEvent } from "../../event_sourcing/ddl_types.ts";
import { isDDLEvent, DDLOperationType } from "../../event_sourcing/ddl_types.ts";
import { createTemplateEvent, validateParams } from "../../event_sourcing/core.ts";
import { TemplateRegistry } from "../../event_sourcing/template_event_store.ts";
import { StateCache } from "../cache/state_cache.ts";
import { SchemaManager } from "../schema_manager.ts";
import { ExtendedTemplateRegistry } from "../../event_sourcing/ddl_event_handler.ts";
import * as telemetry from "../../telemetry_log.ts";

interface PythonResponse {
  success?: boolean;
  error?: string;
  state?: LocalState;
  result?: any;
  message?: string;
  traceback?: string;
}

export class KuzuNativeClientImpl implements KuzuWasmClient {
  private process?: Deno.ChildProcess;
  private writer?: WritableStreamDefaultWriter<string>;
  private reader?: ReadableStreamDefaultReader<Uint8Array>;
  private decoder = new TextDecoder();
  private events: TemplateEvent[] = [];
  private remoteEventHandlers: Array<(event: TemplateEvent) => void> = [];
  private registry = new TemplateRegistry();
  private extendedRegistry = new ExtendedTemplateRegistry();
  private clientId = `native_${crypto.randomUUID()}`;
  private stateCache = new StateCache();
  private schemaManager: SchemaManager;
  private initialized = false;

  constructor() {
    this.schemaManager = new SchemaManager(this.clientId);
  }

  async initialize(): Promise<void> {
    telemetry.info("Initializing native KuzuDB client", {
      clientId: this.clientId
    });

    // Start Python subprocess
    // Use the Python environment from Nix that includes kuzu
    const pythonPath = Deno.env.get("PYTHON_PATH") || "python";
    const scriptPath = new URL("kuzu_native_client.py", import.meta.url).pathname;
    
    telemetry.debug("Starting Python subprocess", {
      pythonPath,
      scriptPath,
      clientId: this.clientId
    });
    
    const command = new Deno.Command(pythonPath, {
      args: [scriptPath],
      stdin: "piped",
      stdout: "piped",
      stderr: "piped",
    });

    this.process = command.spawn();
    
    // Get the writer for stdin as text
    const textEncoder = new TextEncoderStream();
    textEncoder.readable.pipeTo(this.process.stdin);
    this.writer = textEncoder.writable.getWriter();
    
    // Set up stdout reader
    this.reader = this.process.stdout.getReader();
    
    // Set up stderr reader for logging
    const stderrReader = this.process.stderr.getReader();
    this.startStderrReader(stderrReader, this.decoder);
    
    // Send initialization request
    const initRequest = {
      method: "initialize",
      params: {}
    };
    
    await this.sendRequest(initRequest);
    const response = await this.readResponse(this.reader, this.decoder);
    
    if (response.error) {
      throw new Error(`Failed to initialize KuzuDB: ${response.error}`);
    }
    
    this.initialized = true;
    telemetry.info("Native KuzuDB client initialized successfully");
  }
  
  private async sendRequest(request: any): Promise<void> {
    if (!this.writer) {
      throw new Error("Client not initialized");
    }
    
    const data = JSON.stringify(request) + "\n";
    await this.writer.write(data);
  }
  
  private async readResponse(reader: ReadableStreamDefaultReader<Uint8Array>, decoder: TextDecoder): Promise<PythonResponse> {
    let buffer = "";
    
    // Read until we get a complete line
    while (true) {
      const { value, done } = await reader.read();
      
      if (done) {
        throw new Error("Python process terminated unexpectedly");
      }
      
      buffer += decoder.decode(value, { stream: true });
      
      // Check if we have a complete line
      const newlineIndex = buffer.indexOf('\n');
      if (newlineIndex !== -1) {
        const line = buffer.substring(0, newlineIndex).trim();
        
        // Keep remaining data in buffer
        buffer = buffer.substring(newlineIndex + 1);
        
        if (line) {
          try {
            return JSON.parse(line);
          } catch (e) {
            telemetry.error("Failed to parse response", { line, error: String(e) });
            throw new Error(`Invalid JSON response: ${line}`);
          }
        }
      }
    }
  }
  
  private startStderrReader(reader: ReadableStreamDefaultReader<Uint8Array>, decoder: TextDecoder): void {
    const readStderr = async () => {
      try {
        while (true) {
          const { value, done } = await reader.read();
          if (done) break;
          
          const text = decoder.decode(value);
          const lines = text.trim().split('\n');
          
          for (const line of lines) {
            if (line.trim()) {
              try {
                const logEntry = JSON.parse(line);
                // Forward Python logs to our telemetry
                if (logEntry.level && logEntry.message) {
                  telemetry.log(logEntry.level.toLowerCase(), {
                    source: "python_native_client",
                    ...logEntry
                  });
                }
              } catch {
                // Not JSON, just log as debug
                telemetry.debug("Python stderr output", { output: line });
              }
            }
          }
        }
      } catch (error) {
        telemetry.error("Error reading stderr", { error: String(error) });
      }
    };
    
    readStderr();
  }

  async initializeFromSnapshot(snapshot: EventSnapshot): Promise<void> {
    await this.initialize();
    
    // Clear existing state
    this.events = [];
    this.stateCache.clear();
    
    // Initialize schema manager from snapshot
    await this.schemaManager.initializeFromSnapshot(
      snapshot.events,
      (query: string) => this.executeQuery(query)
    );
    
    // Replay all events from snapshot
    telemetry.info("Replaying events from snapshot", { count: snapshot.events.length });
    for (const event of snapshot.events) {
      await this.applyEvent(event);
      this.events.push(event);
    }
  }

  async executeTemplate(template: string, params: Record<string, any>): Promise<TemplateEvent> {
    if (!this.initialized || !this.process) {
      throw new Error("Client not initialized");
    }
    
    // Check if this is a DDL template
    if (this.extendedRegistry.isDDLTemplate(template)) {
      // Create DDL event
      const ddlEvent = this.schemaManager.createDDLEvent(
        template as DDLOperationType,
        params
      );
      
      // Apply DDL event
      await this.applyEvent(ddlEvent);
      
      // Store event
      this.events.push(ddlEvent);
      
      return ddlEvent;
    }
    
    // For DML templates, use extended registry for validation
    const metadata = this.extendedRegistry.getTemplateMetadata(template);
    const sanitizedParams = validateParams(params, metadata);
    
    // Check for injection attempts
    for (const value of Object.values(sanitizedParams)) {
      if (typeof value === 'string' && 
          (value.includes("DROP") || value.includes("DELETE") || value.includes("--"))) {
        throw new Error("Invalid parameter: potential injection attempt");
      }
    }
    
    // Create event
    const event = createTemplateEvent(template, sanitizedParams, this.clientId);
    
    // Execute via Python subprocess
    const request = {
      method: "executeTemplate",
      params: {
        template: template,
        params: sanitizedParams
      }
    };
    
    await this.sendRequest(request);
    const response = await this.readResponse(this.reader!, this.decoder);
    
    if (response.error) {
      throw new Error(`Failed to execute template: ${response.error}`);
    }
    
    // Store event
    this.events.push(event);
    
    // Invalidate cache after applying new event
    this.stateCache.invalidateOnEvent(event);
    
    telemetry.info("Template executed via native client", {
      template,
      eventId: event.id,
      clientId: this.clientId
    });
    
    return event;
  }

  async getLocalState(): Promise<LocalState> {
    if (!this.initialized || !this.process) {
      return { users: [], posts: [], follows: [] };
    }
    
    // Check cache first for O(1) access
    const cachedState = this.stateCache.getCachedState();
    if (cachedState) {
      return cachedState;
    }
    
    try {
      const request = {
        method: "getLocalState",
        params: {}
      };
      
      await this.sendRequest(request);
      const decoder = new TextDecoder();
      const reader = this.process.stdout.getReader();
      const response = await this.readResponse(reader, decoder);
      
      if (response.error) {
        throw new Error(`Failed to get state: ${response.error}`);
      }
      
      const state = response.state || { users: [], posts: [], follows: [] };
      
      // Cache the computed state
      this.stateCache.setCachedState(state);
      
      return state;
    } catch (error) {
      telemetry.error("Error getting local state", { 
        error: String(error),
        clientId: this.clientId 
      });
      return { users: [], posts: [], follows: [] };
    }
  }

  onRemoteEvent(handler: (event: TemplateEvent) => void): void {
    this.remoteEventHandlers.push(handler);
  }

  async executeQuery(cypher: string, params?: Record<string, any>): Promise<any> {
    if (!this.initialized) {
      throw new Error("Client not initialized");
    }
    
    // For DDL queries, execute directly via Python
    // This is used by SchemaManager
    const request = {
      method: "executeQuery",
      params: {
        query: cypher,
        params: params || {}
      }
    };
    
    await this.sendRequest(request);
    const response = await this.readResponse(this.reader!, this.decoder);
    
    if (response.error) {
      throw new Error(`Query execution failed: ${response.error}`);
    }
    
    return response.result;
  }

  async applyEvent(event: TemplateEvent | DDLTemplateEvent): Promise<void> {
    if (!this.initialized || !this.process) {
      throw new Error("Client not initialized");
    }

    // Check if it's a DDL event
    if (isDDLEvent(event)) {
      try {
        await this.schemaManager.applyDDLEvent(
          event,
          (query: string) => this.executeQuery(query)
        );
        // Notify handlers
        this.remoteEventHandlers.forEach(handler => handler(event));
      } catch (error) {
        telemetry.error(`Failed to apply DDL event ${event.id}`, { 
          error: String(error),
          clientId: this.clientId 
        });
        throw error; // DDL errors should propagate
      }
      return;
    }

    // For DML events, execute via Python
    const request = {
      method: "executeTemplate",
      params: {
        template: event.template,
        params: event.params
      }
    };
    
    await this.sendRequest(request);
    const response = await this.readResponse(this.reader!, this.decoder);
    
    if (response.error) {
      telemetry.error("Failed to apply event", {
        eventId: event.id,
        template: event.template,
        error: response.error,
        clientId: this.clientId
      });
      // Continue processing other events
    } else {
      // Invalidate cache since state has changed
      this.stateCache.invalidateOnEvent(event);
      
      // Notify handlers
      this.remoteEventHandlers.forEach(handler => handler(event));
    }
  }
  
  // Query counter value
  async queryCounter(counterId: string): Promise<number> {
    if (!this.initialized || !this.process) {
      throw new Error("Client not initialized");
    }
    
    const request = {
      method: "executeTemplate",
      params: {
        template: "QUERY_COUNTER",
        params: { counterId }
      }
    };
    
    await this.sendRequest(request);
    const response = await this.readResponse(this.reader!, this.decoder);
    
    if (response.error) {
      throw new Error(`Failed to query counter: ${response.error}`);
    }
    
    return response.result || 0;
  }

  // Schema query methods
  getSchemaVersion(): number {
    return this.schemaManager.getSchemaVersion();
  }

  getSchemaState() {
    return this.schemaManager.getCurrentSchema();
  }

  hasTable(tableName: string): boolean {
    return this.schemaManager.hasTable(tableName);
  }

  hasColumn(tableName: string, columnName: string): boolean {
    return this.schemaManager.hasColumn(tableName, columnName);
  }

  getTableSchema(tableName: string) {
    return this.schemaManager.getTableSchema(tableName);
  }

  getSchemaSyncState() {
    return this.schemaManager.getSyncState();
  }

  async resolveSchemaConflict(
    conflictId: string,
    resolution: "APPLY_FIRST" | "APPLY_LAST" | "MANUAL"
  ): Promise<void> {
    await this.schemaManager.resolveConflict(
      conflictId,
      resolution,
      (query: string) => this.executeQuery(query)
    );
  }

  // DDL Event Creation
  createDDLEvent(
    ddlType: DDLOperationType,
    params: Record<string, any>
  ): DDLTemplateEvent {
    return this.schemaManager.createDDLEvent(ddlType, params);
  }

  async applyDDLEvent(event: DDLTemplateEvent): Promise<void> {
    await this.applyEvent(event);
  }

  getAppliedDDLs(): DDLTemplateEvent[] {
    return this.schemaManager.getAppliedDDLs();
  }

  getPendingDDLs(): DDLTemplateEvent[] {
    return this.schemaManager.getPendingDDLs();
  }

  validateDDL(event: DDLTemplateEvent): { valid: boolean; errors: string[] } {
    return this.schemaManager.validateDDL(event);
  }

  // Template Support
  hasTemplate(template: string): boolean {
    return this.extendedRegistry.hasTemplate(template);
  }

  isDDLTemplate(template: string): boolean {
    return this.extendedRegistry.isDDLTemplate(template);
  }

  isDMLTemplate(template: string): boolean {
    return this.extendedRegistry.isDMLTemplate(template);
  }

  getTemplateMetadata(template: string): any {
    return this.extendedRegistry.getTemplateMetadata(template);
  }
  
  // Cleanup
  async close(): Promise<void> {
    if (this.process) {
      try {
        // Close stdin to signal shutdown
        await this.writer?.close();
        
        // Kill the process
        this.process.kill();
        
        // Wait for process to exit
        await this.process.status;
        
        telemetry.info("Native KuzuDB client closed", {
          clientId: this.clientId
        });
      } catch (error) {
        telemetry.error("Error closing native client", {
          error: String(error),
          clientId: this.clientId
        });
      }
    }
  }
}